### ISSUE TITLE
Opacus: DP-SGD with massive epsilon does not attain same accuracy as model with DP-SGD

### ISSUE THREAD

**Issue by RoyRin** [pytorch/opacus#308]
## ❓ Opacus Training models with DP-SGD with a massive epsilon does not attain same accuracy as model with DP-SGD

I trained Resnet9  with **no DP, and achieved : 85.8 % accuracy**
I then retrained Resnet9 with different epsilon for DP-SGD, and got this curve:
![image](https://user-images.githubusercontent.com/12860243/145493464-ea58a071-0d44-4c52-8f46-66ad83ebf3f9.png)

Whereas **epsilon : 1e106  attains accuracy: 76.7 %**

Note: The code I used to generate this plot is available upon request. And If this is not the right place to raise this issue, please let me know.

**Comment by ffuuugor** [comment]
Thanks for raising - this is the right place, as it indeed looks like a bug and worth looking into for us.

Can you share your input privacy params (noise and max_grad_norm)? Often clipping alone could negatively affect the model - e.g. if you set `max_grad_norm=1.` and `noise_multiplier=1e-7`, you'll get horrendous epsilon, but the model can still underperform compared to the original version

**Comment by RoyRin** [comment]
So originally, I tried it with max_grad_norm = 1.2, but then also re-ran it with max-grad-norm = 1.2*400 

![image](https://user-images.githubusercontent.com/12860243/145692660-1fd11c6d-3c17-45b8-b192-3749f37482fd.png)

Which does even worse...
I didn't explicitly adjust the noise_multiplier, so that could be it.

Here is a gist of my code: I trained this on Resnet9, but also validated it on a few one-off examples of Resnet18 (it just takes longer to train resnet18). https://colab.research.google.com/gist/RoyRin/816282e6e173cb35eb3d1df75687053d/building_image_classifier_opacus_issue_shared.ipynb

note: my code isn't the prettiest, but should be pretty clear.  It was hacked together, just to test this issue. It was adapted from the image-classifier-tutorial : https://opacus.ai/tutorials/building_image_classifier .  
(I'd rather share ugly code so that you can look into the issue, sooner, but let me know if parts are unclear.)

**Comment by RoyRin** [comment]
Hmm, perhaps it is because training a model without DP, you iterate over the dataset 1-by-1 (without subsampling), whereas for DP you are doing privacy amplification through subsampling (and you overload the data-loader). This also requires that you do group-norm, rather than batchnorm; which I think is likely to be the issue. As batchnorm seems to do very well for resnet (the model at hand).

I will run an experiment training the model with group norm, and see what the max error is.

Alternatively, it has to do with the overloading of the optimizer.


ref: group-norm vs batch-norm : https://amaarora.github.io/2020/08/09/groupnorm.html

**Comment by RoyRin** [comment]
Hi, 
I wanted to check in here - do you have any idea what may cause this bug? Have you been able to recreate it?

**Comment by ffuuugor** [comment]
Hey @RoyRin 
Sorry for the delay on this.

I've looked into your notebook (thanks for sharing btw) and it seems that the culprit is indeed `GroupNorm`, not opacus. This is also supported by the evidence in the article you linked.

If you replace `BatchNorm` with `GroupNorm` and then train the model without privacy (i.e. not calling `PrivacyEngine`), you'd get the same poor performance you're getting when setting the epsilon really high. Additionally, if later I introduce PrivacyEngine with parameters that effectively disable noise and gradient clipping, the performance doesn't change. 
This suggests, that the drop is accuracy is attributed to `GroupNorm`, not any of the opacus components.
The limitation on using `BatchNorm` is fundamental and not an implementation detail - the nature of the module is incompatible with the concept of DP-SGD (samples are no longer independent and you can't isolate the effect any individual sample have on the overall gradient).

There are, however, good news. `GroupNorm` is not the only option to replace `BatchNorm`. We can also use `InstanceNorm`, and it turns out, that at least in the case of Resnet9 and CIFAR10 it works much better - you get no drop in accuracy at all.  (see notebook for code how to enable this option)

But that's not the whole story. Another reason why your notebook shows unexpected behaviour is `make_private_with_epsilon`. It's based on a binary search for an optimal noise multiplier, and doesn't work well for the extreme values of epsilon. In the last two cells of my notebook I get the same non-trivial noise multipliers for `target_epsilon=1e10` and `target_epsilon=1e20` (which is obviously wrong). To make things worse, when generating noise, `noise_multiplier` is multiplied by `max_grad_norm`, which you set high to avoid gradient clipping, which results in quite significant amount of noise being added. 
This is all to say that `make_private_with_epsilon` doesn't do what's expected of it for extremely large values of epsilon and for these use cases I recommend setting `noise_multiplier` directly.

My notebook with more detailed comments and metrics: [link](https://drive.google.com/file/d/18kUBnMoXxpV-xgwFh4AMdQtMPKlnyLJH/view?usp=sharing)
(Note that I'm using larger GPU, so you might need to adjust `MAX_PHYSICAL_BATCH_SIZE` to run on Colab) 

cc @karthikprasad - We found at least one use case where it's useful to replace `BatchNorm` with `InstanceNorm` instead of `GroupNorm`. Do you think we can make it easier to switch between the two options in `ModelValidator`? The current option is quite cumbersome.

cc @alexandresablayrolles - `make_private_with_epsilon` doesn't work well for extreme values of epsilon, do you think we can pass down some  binary search parameters (like `sigma_precision` or `sigma_max/min`) to make it more flexible?

PS: Oh, and thanks for noticing wrong data loader variable use - you're right, we should be using one returned from the `PrivacyEngine`. Fixing in #338

**Comment by alexandresablayrolles** [comment]
Good point - currently these are set as global variables but I believe they used to be arguments to make_private_with_epsilon. Let's change that back.

**Comment by tholop** [comment]
Just chiming in, since you were discussing the binary search. I think that there is a small bug (or unexpected behavior) where the binary search can return a value of sigma that yields epsilon > target_epsilon. More details in this issue: https://github.com/pytorch/opacus/issues/346.

**Comment by ffuuugor** [comment]
Closing this issue, since
1) We've identified the root cause: it's BatchNorm and unreliable `make_private_with_epsilon`
2) We've addressed the BatchNorm issue in #340 by adding an option to use InstanceNorm instead
3) We've addressed `make_private_with_epsilon` behaviour in #348
